onmt/utils/Cuda.lua

```diff
@@ -52,16 +52,21 @@ end
   When using CPU only, converts to float instead of the default double.
 ]]
 function Cuda.convert(obj)
-  if torch.typename(obj) then
+  local objtype = torch.typename(obj)
+  if objtype then
     if Cuda.activated and obj.cuda ~= nil then
-      return obj:cuda()
-    elseif not Cuda.activated and obj.float ~= nil then
+      if objtype:find('torch%..*LongTensor') then
+        return obj:cudaLong()
+      else
+        return obj:cuda()
+      end
+    elseif not Cuda.activated and obj.float ~= nil and not objtype:find('torch%..*LongTensor') then
```

Comment:
I guess you can, it's just that your code now converts a `CudaLongTensor` back to a `FloatTensor`. With this piece:

```lua
if objtype:find('torch%..*LongTensor') then
  return obj:long()
else
  return obj:float()
end
```

there is a nice symmetry with the GPU part and works as expected. By the way, calling `long()` on a `LongTensor` does nothing.